internal/auth/oidc/repository_auth_method_update.go

```diff
@@ -17,32 +31,561 @@ import "context"
+			cp.DiscoveryUrl = new.DiscoveryUrl
+		case "ClientId":
+			cp.ClientId = new.ClientId
+		case "ClientSecret":
+			cp.ClientSecret = new.ClientSecret
+		case "MaxAge":
+			cp.MaxAge = new.MaxAge
+		case "SigningAlgs":
+			switch {
+			case len(new.SigningAlgs) == 0:
+				cp.SigningAlgs = nil
+			default:
+				cp.SigningAlgs = make([]string, 0, len(new.SigningAlgs))
+				cp.SigningAlgs = append(cp.SigningAlgs, new.SigningAlgs...)
+			}
+		case "CallbackUrls":
+			switch {
+			case len(new.CallbackUrls) == 0:
+				cp.CallbackUrls = nil
+			default:
+				cp.CallbackUrls = make([]string, 0, len(new.CallbackUrls))
+				cp.CallbackUrls = append(cp.CallbackUrls, new.CallbackUrls...)
+			}
+		case "AudClaims":
+			switch {
+			case len(new.AudClaims) == 0:
+				cp.AudClaims = nil
+			default:
+				cp.AudClaims = make([]string, 0, len(new.AudClaims))
+				cp.AudClaims = append(cp.AudClaims, new.AudClaims...)
+			}
+		case "Certificates":
+			switch {
+			case len(new.Certificates) == 0:
+				cp.Certificates = nil
+			default:
+				cp.Certificates = make([]string, 0, len(new.Certificates))
+				cp.Certificates = append(cp.Certificates, new.Certificates...)
+			}
+		}
+	}
+	return cp
+}
+
+// ValidateAuthMethod will test/validate the provided AuthMethod.
 //
 // It will verify that all required fields for a working AuthMethod have values.
 //
-// If the AuthMethod contains a DiscoveryUrl for an OIDC provider, TestAuthMethod
+// If the AuthMethod contains a DiscoveryUrl for an OIDC provider, ValidateAuthMethod
 // retrieves the OpenID Configuration document. The values in the AuthMethod
 // (and associated data) are validated with the retrieved document. The issuer and
 // id token signing algorithm in the configuration are validated with the
-// retrieved document. TestAuthMethod also verifies the authorization, token,
+// retrieved document. ValidateAuthMethod also verifies the authorization, token,
 // and user_info endpoints by connecting to each and uses any certificates in the
 // configuration as trust anchors to confirm connectivity.
 //
 // Options supported are: WithPublicId, WithAuthMethod
-func (r *Repository) TestAuthMethod(ctx context.Context, opt ...Option) error {
-	panic("to-do")
+func (r *Repository) ValidateAuthMethod(ctx context.Context, opt ...Option) error {
+	const op = "oidc.(Repository).TestAuthMethod"
+	opts := getOpts(opt...)
+	var am *AuthMethod
+	switch {
+	case opts.withPublicId != "":
+		var err error
+		am, err = r.lookupAuthMethod(ctx, opts.withPublicId)
+		if err != nil {
+			return errors.Wrap(err, op)
+		}
+	case opts.withAuthMethod != nil:
+		am = opts.withAuthMethod
+	default:
+		return errors.New(errors.InvalidParameter, op, "neither WithPublicId(...) nor WithAuthMethod(...) options were provided")
+	}
+
+	if err := am.isComplete(); err != nil {
+		return errors.Wrap(err, op)
+	}
+
+	// FYI: once converted to an oidc.Provider, any certs configured will be used as trust anchors for all HTTP requests
```

Comment:
Can you explain this? To me this sounds like you are using (and modifying) `http.DefaultClient`/`http.DefaultTransport`. If so -- don't do that. Use `go-cleanhttp` instead. If it just means "for the requests in this function" then that's fine.

Reply:
`convertToProvider(...)` will convert the AuthMethod to a CAP `oidc.Provider`.  Internally, the cap provider uses go-cleanhttp to get a pooled transport and then if the provider was configured with certs, it creates a cert pool for the http.Client.  Here's the code in cap: https://github.com/hashicorp/cap/blob/4b74380f71bda210a64705240c21c6e309f17039/oidc/provider.go#L576

Reply:
All this results in `provider.DiscoveryInfo(ctx)` will use the certs configured for the auth method.